pkg/cmd/p2p/promote/p2p_promote.go

```diff
@@ -0,0 +1,183 @@
+package promote
+
+import (
+	"fmt"
+	"github.com/coreeng/corectl/pkg/cmdutil/userio"
+	"github.com/spf13/cobra"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+type promoteOpts struct {
+	ImageWithTag       string
+	SourceRegistry     string
+	SourceStage        string
+	SourceAuthOverride string
+	DestRegistry       string
+	DestStage          string
+	DestAuthOverride   string
+	Streams            userio.IOStreams
+}
+
+type imageOpts struct {
+	ImageNameWithTag string
+	Registry         string
+	RepoPath         string
+	AuthOverride     string
+}
+
+func NewP2PPromoteCmd() (*cobra.Command, error) {
+	var opts = promoteOpts{}
+	var promoteCommand = &cobra.Command{
+		Use:   "promote <image_with_tag>",
+		Short: "Promotes image",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts.ImageWithTag = args[0]
+
+			opts.Streams = userio.NewIOStreams(
+				cmd.InOrStdin(),
+				cmd.OutOrStdout(),
+			)
+
+			return run(&opts)
+		},
+	}
+
+	requiredFlagMap := map[string]*string{
+		"source-registry": &opts.SourceRegistry,
+		"source-stage":    &opts.SourceStage,
+		"dest-registry":   &opts.DestRegistry,
+		"dest-stage":      &opts.DestStage,
+	}
+
+	for name, field := range requiredFlagMap {
+		err := addFlag(promoteCommand, field, name, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	optionalFlagMap := map[string]*string{
+		"source-auth-override": &opts.SourceAuthOverride,
+		"dest-auth-override":   &opts.DestAuthOverride,
+	}
+	for name, field := range optionalFlagMap {
+		err := addFlag(promoteCommand, field, name, false)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return promoteCommand, nil
+}
+
+func addFlag(promoteCommand *cobra.Command, field *string, name string, required bool) error {
+	envVariableName := strings.ToUpper(strings.ReplaceAll(name, "-", "_"))
+
+	description := fmt.Sprintf("optional, defaults to environment variable: %s", envVariableName)
+	if required {
+		description = fmt.Sprintf("required, defaults to environment variable: %s", envVariableName)
+	}
+
+	promoteCommand.Flags().StringVar(
+		field,
+		name,
+		"",
+		description,
+	)
+
+	envVariableValue := os.Getenv(envVariableName)
+
+	if envVariableValue != "" {
+		*field = envVariableValue
+	}
+
+	if envVariableValue == "" && required {
+		if err := promoteCommand.MarkFlagRequired(name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func run(opts *promoteOpts) error {
+
+	sourceImage := &imageOpts{
+		ImageNameWithTag: opts.ImageWithTag,
+		Registry:         opts.SourceRegistry,
+		RepoPath:         opts.SourceStage,
+		AuthOverride:     opts.SourceAuthOverride,
+	}
+
+	destinationImage := &imageOpts{
+		ImageNameWithTag: opts.ImageWithTag,
+		Registry:         opts.DestRegistry,
+		RepoPath:         opts.DestStage,
+		AuthOverride:     opts.DestAuthOverride,
+	}
+
+	logInfo := opts.Streams.Info
+	logInfo("Configuring docker with gcloud")
+	output, err := configureDockerWithGcloud()
```

Comment:
Should we just connect the current process stdout+stderr to the created subprocesses?
I guess we don't need formatting the Info logging does

Reply:
not sure what you meant exactly but I've used cobras `cobra.Command.OutOrStdout` as a writer

Reply:
I mean, when you create a command with `exec` package, you can specify the Writer for both: Stdout and Stderr:
```go
	command := exec.Command(cmd, args...)
	command.Stdout = os.Stdout
	command.Stderr = os.Stderr
```